File: app/src/main/java/org/y20k/trackbook/helpers/LocationHelper.java
/**
 * LocationHelper.java
 * Implements the LocationHelper class
 * A LocationHelper offers helper methods for dealing with location issues
 *
 * This file is part of
 * TRACKBOOK - Movement Recorder for Android
 *
 * Copyright (c) 2016-18 - Y20K.org
 * Licensed under the MIT-License
 * http://opensource.org/licenses/MIT
 *
 * Trackbook uses osmdroid - OpenStreetMap-Tools for Android
 * https://github.com/osmdroid/osmdroid
 */

package org.y20k.trackbook.helpers;

import android.content.Context;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.SystemClock;
import android.provider.Settings;

import java.util.List;
import java.util.Locale;
import java.util.concurrent.TimeUnit;

import androidx.annotation.Nullable;


/**
 * LocationHelper class
 */
public final class LocationHelper implements TrackbookKeys {

    /* Define log tag */
    private static final String LOG_TAG = LocationHelper.class.getSimpleName();


    /* Determines last known location  */
    public static Location determineLastKnownLocation(LocationManager locationManager) {
        // define variables
        List locationProviders = locationManager.getProviders(true);
        Location gpsLocation = null;
        Location networkLocation = null;

        // set location providers
        String gpsProvider = LocationManager.GPS_PROVIDER;
        String networkProvider = LocationManager.NETWORK_PROVIDER;


        if (locationProviders.contains(gpsProvider)) {
            // get last know location from gps
            try {
                gpsLocation = locationManager.getLastKnownLocation(gpsProvider);
            } catch (SecurityException e) {
                // catches permission problems
                e.printStackTrace();
            }
        }

        if (locationProviders.contains(networkProvider)) {
            // get last known location from wifi and cell
            try {
                networkLocation = locationManager.getLastKnownLocation(networkProvider);
            } catch (SecurityException e) {
                // catches permission problems
                e.printStackTrace();
            }
        }

        if (gpsLocation == null) {
            return networkLocation;
        } else if (networkLocation == null) {
            return gpsLocation;
        } else if (isBetterLocation(gpsLocation, networkLocation)) {
            return gpsLocation;
        } else {
            return networkLocation;
        }
    }


    /* Determines whether one location reading is better than the current location fix */
    public static boolean isBetterLocation(Location location, Location currentBestLocation) {
        // credit: the isBetterLocation method was sample code from: https://developer.android.com/guide/topics/location/strategies.html

        if (currentBestLocation == null) {
            // a new location is always better than no location
            return true;
        }

        // check whether the new location fix is newer or older
        long timeDelta = location.getElapsedRealtimeNanos() - currentBestLocation.getElapsedRealtimeNanos();
        boolean isSignificantlyNewer = timeDelta > TWO_MINUTES_IN_NANOSECONDS;
        boolean isSignificantlyOlder = timeDelta < -TWO_MINUTES_IN_NANOSECONDS;
        boolean isNewer = timeDelta > 0;

        // if it's been more than two minutes since the current location, use the new location because the user has likely moved
        if (isSignificantlyNewer) {
            return true;
        } else if (isSignificantlyOlder) {
            return false;
        }

        // check whether the new location fix is more or less accurate
        int accuracyDelta = (int) (location.getAccuracy() - currentBestLocation.getAccuracy());
        boolean isLessAccurate = accuracyDelta > 0;
        boolean isMoreAccurate = accuracyDelta < 0;
        boolean isSignificantlyLessAccurate = accuracyDelta > 200;

        // check if the old and new location are from the same provider
        boolean isFromSameProvider = isSameProvider(location.getProvider(), currentBestLocation.getProvider());

        // determine location quality using a combination of timeliness and accuracy
        if (isMoreAccurate) {
            return true;
        } else if (isNewer && !isLessAccurate) {
            return true;
        } else if (isNewer && !isSignificantlyLessAccurate && isFromSameProvider) {
            return true;
        }
        return false;
    }


    /* Checks accuracy of given location */
    public static boolean isAccurate(Location location) {
        return location.getAccuracy() < FIFTY_METER_RADIUS;
    }


    /* Checks if given location is newer than two minutes */
    public static boolean isCurrent(Location location) {
        if (location == null) {
            return false;
        } else {
            long locationAge = SystemClock.elapsedRealtimeNanos() - location.getElapsedRealtimeNanos();
            return locationAge < TWO_MINUTES_IN_NANOSECONDS;
        }
    }


    /* Checks if given location is a new WayPoint */
    public static boolean isNewWayPoint(Location lastLocation, Location newLocation, float averageSpeed) {
        float distance = newLocation.distanceTo(lastLocation);
        long timeDifference = newLocation.getElapsedRealtimeNanos() - lastLocation.getElapsedRealtimeNanos();

        if (newLocation.getProvider().equals(LocationManager.NETWORK_PROVIDER)) {
            // calculate speed difference
            float speedDifference;
            float currentSpeed = distance / ((float)timeDifference / ONE_SECOND_IN_NANOSECOND);
            if (currentSpeed > averageSpeed) {
                speedDifference = currentSpeed / averageSpeed;
            } else {
                speedDifference = averageSpeed / currentSpeed;
            }

            // SPECIAL CASE network: plausibility check for network provider. looking for sudden location jump errors
            if (averageSpeed != 0f && currentSpeed > 10f && speedDifference > 2f) {
                // implausible location (speed is high (10 m/s == 36km/h) and has doubled)
                return false;
            }

            // SPECIAL CASE network: if last location came from gps. only accept location fixes with decent accuracy
            if (lastLocation.getProvider().equals(LocationManager.GPS_PROVIDER) && newLocation.getAccuracy() < 66) {
                // network locations tend to be too in accurate
                return false;
            }

            // DEFAULT network: distance is bigger than 30 meters and time difference bigger than 12 seconds
            return distance > 30 && timeDifference >= 12 * ONE_SECOND_IN_NANOSECOND; // TODO add minimal accuracy

        } else {
            // DEFAULT GPS: distance is bigger than 10 meters and time difference bigger than 12 seconds
            return distance > 10 && timeDifference >= 12 * ONE_SECOND_IN_NANOSECOND;
        }

    }


    /* Checks if given location is a stop over */
    public static boolean isStopOver(@Nullable Location previousLocation, Location newLocation) {
        if (previousLocation != null) {
            long timeDifference =  newLocation.getElapsedRealtimeNanos() - previousLocation.getElapsedRealtimeNanos();
            return timeDifference >= FIVE_MINUTES_IN_NANOSECONDS;
        } else {
            return false;
        }
    }


    /* Registers gps and network location listeners */
    public static void registerLocationListeners(LocationManager locationManager, LocationListener gpsListener, LocationListener networkListener) {
        LogHelper.v(LOG_TAG, "Registering location listeners.");

        // get location providers
        List locationProviders = locationManager.getAllProviders();

        // got GPS location provider?
        if (gpsListener != null && locationProviders.contains(LocationManager.GPS_PROVIDER)) {
            try {
                // register GPS location listener and request updates
                locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 0, 0, gpsListener);
                LogHelper.v(LOG_TAG, "Registering gps listener.");
            } catch (SecurityException e) {
                // catches permission problems
                e.printStackTrace();
            }
        }

        // got network location provider?
        if (networkListener != null && locationProviders.contains(LocationManager.NETWORK_PROVIDER)) {
            try {
                // register network location listener and request updates
                locationManager.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, 0, 0, networkListener);
                LogHelper.v(LOG_TAG, "Registering network listener.");
            } catch (SecurityException e) {
                // catches permission problems
                e.printStackTrace();
            }
        }
    }


    /* Removes gps and network location listeners */
    public static void removeLocationListeners(LocationManager locationManager, LocationListener gpsListener, LocationListener networkListener) {
        LogHelper.v(LOG_TAG, "Removing location listeners.");

        // get location providers
        List locationProviders = locationManager.getAllProviders();

        // got GPS location provider?
        if (locationProviders.contains(LocationManager.GPS_PROVIDER) && gpsListener != null) {
            try {
                // remove GPS listener
                locationManager.removeUpdates(gpsListener);
                LogHelper.v(LOG_TAG, "Removing gps listener.");
            } catch (SecurityException e) {
                // catches permission problems
                e.printStackTrace();
            }
        }

        // got network location provider?
        if (locationProviders.contains(LocationManager.NETWORK_PROVIDER) && networkListener != null) {
            try {
                // remove network listener
                locationManager.removeUpdates(networkListener);
                LogHelper.v(LOG_TAG, "Removing network listener.");
            } catch (SecurityException e) {
                // catches permission problems
                e.printStackTrace();
            }
        }

    }


    /* Converts milliseconds to mm:ss or hh:mm:ss */
    public static String convertToReadableTime(long milliseconds, boolean includeHours) {

        if (includeHours) {
            // format hh:mm:ss
            return String.format(Locale.ENGLISH, "%02d:%02d:%02d", TimeUnit.MILLISECONDS.toHours(milliseconds),
                    TimeUnit.MILLISECONDS.toMinutes(milliseconds) % TimeUnit.HOURS.toMinutes(1),
                    TimeUnit.MILLISECONDS.toSeconds(milliseconds) % TimeUnit.MINUTES.toSeconds(1));
        } else if (TimeUnit.MILLISECONDS.toHours(milliseconds) < 1) {
            // format mm:ss
            return String.format(Locale.ENGLISH, "%02d:%02d", TimeUnit.MILLISECONDS.toMinutes(milliseconds) % TimeUnit.HOURS.toMinutes(1),
                    TimeUnit.MILLISECONDS.toSeconds(milliseconds) % TimeUnit.MINUTES.toSeconds(1));
        } else {
            return null;
        }

    }


    /* Check if any location provider is enabled */
    public static boolean checkLocationSystemSetting(Context context) {
        int locationSettingState = 0;
        try {
            locationSettingState = Settings.Secure.getInt(context.getContentResolver(), Settings.Secure.LOCATION_MODE);
        } catch (Settings.SettingNotFoundException e) {
            e.printStackTrace();
        }
        return locationSettingState != Settings.Secure.LOCATION_MODE_OFF;
    }


    /* Checks whether two location providers are the same */
    private static boolean isSameProvider(String provider1, String provider2) {
        // credit: the isSameProvider method was sample code from: https://developer.android.com/guide/topics/location/strategies.html
        if (provider1 == null) {
            return provider2 == null;
        }
        return provider1.equals(provider2);
    }

}

File: app/src/main/java/org/y20k/trackbook/helpers/LocationHelper.kt
/*
 * LocationHelper.kt
 * Implements the LocationHelper object
 * A LocationHelper offers helper methods for dealing with location issues
 *
 * This file is part of
 * TRACKBOOK - Movement Recorder for Android
 *
 * Copyright (c) 2016-20 - Y20K.org
 * Licensed under the MIT-License
 * http://opensource.org/licenses/MIT
 *
 * Trackbook uses osmdroid - OpenStreetMap-Tools for Android
 * https://github.com/osmdroid/osmdroid
 */


package org.y20k.trackbook.helpers

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationManager
import android.os.SystemClock
import androidx.core.content.ContextCompat
import org.y20k.trackbook.Keys
import org.y20k.trackbook.core.Track
import org.y20k.trackbook.core.WayPoint
import java.util.*
import kotlin.math.pow


/*
 * Keys object
 */
object LocationHelper {

    /* Define log tag */
    private val TAG: String = LogHelper.makeLogTag(LocationHelper::class.java)


    /* Get default location */
    fun getDefaultLocation(): Location {
        val defaultLocation: Location = Location(LocationManager.NETWORK_PROVIDER)
        defaultLocation.latitude = Keys.DEFAULT_LATITUDE
        defaultLocation.longitude = Keys.DEFAULT_LONGITUDE
        defaultLocation.accuracy = Keys.DEFAULT_ACCURACY
        defaultLocation.altitude = Keys.DEFAULT_ALTITUDE
        defaultLocation.time = Keys.DEFAULT_DATE.time
        return defaultLocation
    }


    /* Checks if a location is older than one minute */
    fun isOldLocation(location: Location): Boolean {
        // check how many milliseconds the given location is old
        return GregorianCalendar.getInstance().time.time - location.time > Keys.SIGNIFICANT_TIME_DIFFERENCE
    }


    /* Tries to return the last location that the system has stored */
    fun getLastKnownLocation(context: Context): Location {
        // get last location that Trackbook has stored
        var lastKnownLocation: Location = PreferencesHelper.loadCurrentBestLocation(context)
        // try to get the last location the system has stored - it is probably more recent
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
            val locationManager = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager
            val lastKnownLocationGps: Location = locationManager.getLastKnownLocation(LocationManager.GPS_PROVIDER) ?: lastKnownLocation
            val lastKnownLocationNetwork: Location = locationManager.getLastKnownLocation(LocationManager.NETWORK_PROVIDER) ?: lastKnownLocation
            when (isBetterLocation(lastKnownLocationGps, lastKnownLocationNetwork)) {
                true -> lastKnownLocation = lastKnownLocationGps
                false -> lastKnownLocation = lastKnownLocationNetwork
            }
        }
        return lastKnownLocation
    }


    /* Determines whether one location reading is better than the current location fix */
    fun isBetterLocation(location: Location, currentBestLocation: Location?): Boolean {
        // Credit: https://developer.android.com/guide/topics/location/strategies.html#BestEstimate

        if (currentBestLocation == null) {
            // a new location is always better than no location
            return true
        }

        // check whether the new location fix is newer or older
        val timeDelta: Long = location.time - currentBestLocation.time
        val isSignificantlyNewer: Boolean = timeDelta > Keys.SIGNIFICANT_TIME_DIFFERENCE
        val isSignificantlyOlder:Boolean = timeDelta < -Keys.SIGNIFICANT_TIME_DIFFERENCE

        when {
            // if it's been more than two minutes since the current location, use the new location because the user has likely moved
            isSignificantlyNewer -> return true
            // if the new location is more than two minutes older, it must be worse
            isSignificantlyOlder -> return false
        }

        // check whether the new location fix is more or less accurate
        val isNewer: Boolean = timeDelta > 0L
        val accuracyDelta: Float = location.accuracy - currentBestLocation.accuracy
        val isLessAccurate: Boolean = accuracyDelta > 0f
        val isMoreAccurate: Boolean = accuracyDelta < 0f
        val isSignificantlyLessAccurate: Boolean = accuracyDelta > 200f

        // check if the old and new location are from the same provider
        val isFromSameProvider: Boolean = location.provider == currentBestLocation.provider

        // determine location quality using a combination of timeliness and accuracy
        return when {
            isMoreAccurate -> true
            isNewer && !isLessAccurate -> true
            isNewer && !isSignificantlyLessAccurate && isFromSameProvider -> true
            else -> false
        }
    }


    /* Checks if GPS location provider is available and enabled */
    fun isGpsEnabled(locationManager: LocationManager): Boolean {
        if (locationManager.allProviders.contains(LocationManager.GPS_PROVIDER)) {
            return locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)
        } else {
            return false
        }
    }


    /* Checks if Network location provider is available and enabled */
    fun isNetworkEnabled(locationManager: LocationManager): Boolean {
        if (locationManager.allProviders.contains(LocationManager.NETWORK_PROVIDER)) {
            return locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER)
        } else {
            return false
        }
    }



    /* Checks if given location is new */
    fun isRecentEnough(location: Location): Boolean {
        val locationAge: Long = SystemClock.elapsedRealtimeNanos() - location.elapsedRealtimeNanos
        return locationAge < Keys.DEFAULT_THRESHOLD_LOCATION_AGE
    }


    /* Checks if given location is accurate */
    fun isAccurateEnough(location: Location, locationAccuracyThreshold: Int): Boolean {
        val isAccurate: Boolean
        when (location.provider) {
            LocationManager.GPS_PROVIDER -> isAccurate = location.accuracy < locationAccuracyThreshold
            else -> isAccurate = location.accuracy < locationAccuracyThreshold + 10 // a bit more relaxed when location comes from network provider
        }
        return isAccurate
    }


    /* Checks if the first location of track is plausible */
    fun isFirstLocationPlausible(secondLocation: Location, track: Track): Boolean {
        // speed in km/h
        val speed: Double = calculateSpeed(firstLocation = track.wayPoints[0].toLocation(), secondLocation = secondLocation, firstTimestamp = track.recordingStart.time, secondTimestamp = GregorianCalendar.getInstance().time.time)
        // plausible = speed under 250 km/h
        return speed < Keys.IMPLAUSIBLE_TRACK_START_SPEED
    }


    /* Calculates speed */
    private fun calculateSpeed(firstLocation: Location, secondLocation: Location, firstTimestamp: Long, secondTimestamp: Long, useImperial: Boolean = false): Double {
        // time difference in seconds
        val timeDifference: Long = (secondTimestamp - firstTimestamp) / 1000L
        // distance in meters
        val distance: Float = calculateDistance(firstLocation, secondLocation)
        // speed in either km/h (default) or mph
        return LengthUnitHelper.convertMetersPerSecond(distance / timeDifference, useImperial)
    }


    /* Checks if given location is different enough compared to previous location */
    fun isDifferentEnough(previousLocation: Location?, location: Location, accuracyMultiplier: Int): Boolean {
        // check if previous location is (not) available
        if (previousLocation == null) return true

        // location.accuracy is given as 1 standard deviation, with a 68% chance
        // that the true position is within a circle of this radius.
        // These formulas determine if the difference between the last point and
        // new point is statistically significant.
        val accuracy: Float = if (location.accuracy != 0.0f) location.accuracy else Keys.DEFAULT_THRESHOLD_DISTANCE
        val previousAccuracy: Float = if (previousLocation.accuracy != 0.0f) previousLocation.accuracy else Keys.DEFAULT_THRESHOLD_DISTANCE
        val accuracyDelta: Double = Math.sqrt((accuracy.pow(2) + previousAccuracy.pow(2)).toDouble())
        val distance: Float = calculateDistance(previousLocation, location)

        // With 1*accuracyDelta we have 68% confidence that the points are
        // different. We can multiply this number to increase confidence but
        // decrease point recording frequency if needed.
        return distance > accuracyDelta * accuracyMultiplier
    }


    /* Calculates distance in meters between two locations */
    fun calculateDistance(previousLocation: Location?, location: Location): Float  {
        var distance: Float = 0f
        // two data points needed to calculate distance
        if (previousLocation != null) {
            // add up distance
            distance = previousLocation.distanceTo(location)
        }
        return distance
    }


    /* Calculate elevation differences */
    fun calculateElevationDifferencesOld(previousLocation: Location?, location: Location, track: Track): Pair<Double, Double> {
        // store current values
        var positiveElevation: Double = track.positiveElevation
        var negativeElevation: Double = track.negativeElevation
        if (previousLocation != null) {
            // factor is bigger than 1 if the time stamp difference is larger than the movement recording interval
            val timeDifferenceFactor: Long = (location.time - previousLocation.time) / Keys.ADD_WAYPOINT_TO_TRACK_INTERVAL
            // get elevation difference and sum it up
            val altitudeDifference: Double = location.altitude - previousLocation.altitude
            if (altitudeDifference > 0 && altitudeDifference < Keys.ALTITUDE_MEASUREMENT_ERROR_THRESHOLD * timeDifferenceFactor && location.altitude != Keys.DEFAULT_ALTITUDE) {
                positiveElevation = track.positiveElevation + altitudeDifference // upwards movement
            }
            if (altitudeDifference < 0 && altitudeDifference > -Keys.ALTITUDE_MEASUREMENT_ERROR_THRESHOLD * timeDifferenceFactor && location.altitude != Keys.DEFAULT_ALTITUDE) {
                negativeElevation = track.negativeElevation + altitudeDifference // downwards movement
            }
        }
        return Pair(positiveElevation, negativeElevation)
    }


    /* Calculate elevation differences */
    fun calculateElevationDifferences(previousLocation: Location?, location: Location, track: Track, altitudeSmoothingValue: Int): Pair<Double, Double> {
        // store current values
        var positiveElevation: Double = track.positiveElevation
        var negativeElevation: Double = track.negativeElevation
        if (previousLocation != null && location.altitude != Keys.DEFAULT_ALTITUDE) {
            val locationAltitudeCorrected: Double = calculateCorrectedAltitude(location, track, altitudeSmoothingValue)
            val previousLocationAltitudeCorrected: Double = calculateCorrectedAltitude(previousLocation, track, altitudeSmoothingValue)
            // get elevation difference and sum it up
            val altitudeDifference: Double = locationAltitudeCorrected - previousLocationAltitudeCorrected
            if (altitudeDifference > 0) {
                positiveElevation = track.positiveElevation + altitudeDifference // upwards movement
            }
            if (altitudeDifference < 0) {
                negativeElevation = track.negativeElevation + altitudeDifference // downwards movement
            }
        }
        return Pair(positiveElevation, negativeElevation)
    }


    /* Checks if given location is a stop over */
    fun isStopOver(previousLocation: Location?, location: Location): Boolean {
        if (previousLocation == null) return false
        // check how many milliseconds the given locations are apart
        return location.time - previousLocation.time > Keys.STOP_OVER_THRESHOLD
    }


    /* Calculate a moving average taking into account previously recorded altitude values */
    private fun calculateCorrectedAltitude(location: Location, track: Track, altitudeSmoothingValue: Int): Double {
        // add location to track
        track.wayPoints.add(WayPoint(location))
        // get size of track
        val trackSize: Int = track.wayPoints.size
        // skip calculation if less than two waypoints available
        if (trackSize < 2) return location.altitude
        // get number of locations to be used in calculating the moving average
        val numberOfLocationsUsedForSmoothing: Int = if (trackSize < altitudeSmoothingValue) {
            trackSize
        } else {
            altitudeSmoothingValue
        }
        // add altitude values in range and calculate average
        val mostRecentWaypointIndex: Int = trackSize - 1
        var altitudeSum: Double = 0.0
        for (i in mostRecentWaypointIndex..(mostRecentWaypointIndex - numberOfLocationsUsedForSmoothing)) {
            altitudeSum = altitudeSum + track.wayPoints[i].altitude
        }
        return altitudeSum / numberOfLocationsUsedForSmoothing
    }



}
